stdlib/public/TensorFlow/CompilerRuntime.swift

```diff
@@ -1197,6 +1197,150 @@ public func _TFCCreateTensorHandleFromC(
+    buffer.withMemoryRebound(to: TF_DataType.self) { reboundBuffer in
+      TFE_OpSetAttrTypeList(op, attrName, reboundBuffer.baseAddress,
+                            Int32(reboundBuffer.count))
+    }
+  }
+}
+
+@usableFromInline
+@_silgen_name("_swift_tfc_OpSetAttrTensorShapeArray")
+func _TFCOpSetAttrTensorShapeArray(_ op: OpaquePointer,
+                             _ attrName: UnsafePointer<Int8>,
+                             _ value: Array<TensorShape>,
+                             _ status: CTFStatus) {
+  let flattenedDims = value.flatMap { $0.dimensions.map(Int64.init) }
+  let ranks = value.map { $0.rank }
+  setAttrShapeList(op: op, attrName: attrName, flattenedDims: flattenedDims,
+                   ranks: ranks, status: status)
+}
+
+@usableFromInline
+@_silgen_name("_swift_tfc_OpSetAttrOptionalTensorShapeArray")
+func _TFCOpSetAttrOptionalTensorShapeArray(_ op: OpaquePointer,
+                             _ attrName: UnsafePointer<Int8>,
+                             _ value: Array<TensorShape?>,
+                             _ status: CTFStatus) {
+  let flattenedDims = value.flatMap { (tensorShapeOpt) -> [Int64] in
+    if let tensorShape = tensorShapeOpt {
+      return tensorShape.dimensions.map(Int64.init)
+    }
+    return []
+  }
+  let ranks = value.map { tensorShapeOpt -> Int32 in
+    if let tensorShape = tensorShapeOpt {
+      return tensorShape.rank
+    }
+    return -1
+  }
+  setAttrShapeList(op: op, attrName: attrName, flattenedDims: flattenedDims,
+                   ranks: ranks, status: status)
+}
+
+/// Given dimensions and ranks in the form described below, makes the
+/// appropriate call to TFE_OpSetAttrShapeList(op, attrName, ..., status).
+///
+/// - Parameters
+///   - flattenedDims: all the shapes' dimensions concatenated together in
+///     order
+///   - ranks: all the shapes' ranks (-1 denotes unknown rank)
+func setAttrShapeList(op: OpaquePointer, attrName: UnsafePointer<Int8>,
+                      flattenedDims: Array<Int64>, ranks: Array<Int32>,
+                      status: CTFStatus) {
+  flattenedDims.withUnsafeBufferPointer { flattenedDimsBuffer in
+    var dimsPtr: UnsafePointer<Int64>? = flattenedDimsBuffer.baseAddress
+    var dims: [UnsafePointer<Int64>?] = []
+    for rank in ranks {
+      dims.append(dimsPtr)
+      if rank >= 0 {
+        dimsPtr = dimsPtr.map { $0.advanced(by: Int(rank)) }
+      }
+    }
+    dims.withUnsafeMutableBufferPointer { dimsBuffer in
+      ranks.withUnsafeBufferPointer { ranksBuffer in
+        TFE_OpSetAttrShapeList(op, attrName, dimsBuffer.baseAddress,
+                               ranksBuffer.baseAddress,
+                               Int32(ranksBuffer.count), status)
+      }
+    }
+  }
+}
+
+/// Wrapper around TFE_OpSetAttrString that handles converting the Swift Stdlib
+/// String into a buffer that TFE_OpSetAttrString can read.
+@usableFromInline
+@_silgen_name("_swift_tfc_OpSetAttrString")
+func _TFCOpSetAttrString(_ op: OpaquePointer,
```

Comment:
the runtime def func signature has a different set of args:
```
FUNCTION(TFC_OpSetAttrString, _swift_tfc_OpSetAttrString, C_CC,
         RETURNS(),
         ARGS(Int8PtrTy, Int8PtrTy, Int8PtrTy, Int64Ty),
         ATTRS(NoUnwind))
```

I thought we cannot have a runtime function that takes an arg of type "String". The arg/return types have to be more plain types. Did I miss something?

Reply:
What's happening is that "String" is a struct of `i8*, i64`, so Swift "String" values get "exploded" into those two llvm ir values. (I think that's where the word "Explosion" comes from).

Any time there is a swift "String" we can ask for the two llvm ir values (like I do in L2152 of IRGenSIL.cpp), and any time something needs a "String" (like this function) it is actually wanting the two llvm ir values. So it ends up actually passing the String to the runtime function.

It is a bit bad that I have to encode this internal information about String into RuntimeFunctions.def, but that's the only simple way I could think of to pass a String to the runtime.

Reply:
Thanks for the context. It's very interesting and creative that you found this way! But I agree using such internal knowledge is not ideal and could be brittle.

Consider enhancing the comments here and/or at at irgen when you call `IGM.getTFC_OpSetAttrStringFn()` to describe this. 

What could reduce the amount of magic is that the swift function `_TFCOpSetAttrString()` still takes `(void* or char*, int64)` as the last two params, instead of the String param. In the swift type system, I think void* or char* would be a `UnsafePointer<Int8>`. This way the magic and internal knowledge is only limited to the irgen code that expodes a string to these two param values, but not how a runtime function def gets "magically" mapped to a swift runtime function with a different signature.

Reply:
Also to write up some other half-baked ideas (they add more complexity and probably is not a win for now): Based on the fact that a swift `String` can be automatically converted to a `UnsafePointer<UInt8>` value when we call a c function expecting a `const char *` param (https://stackoverflow.com/questions/27063569/string-value-to-unsafepointeruint8-function-parameter-behavior), when we type check an expr like
```swift
#tfop("Conv2D", convImage, convFilter,
                                    T$dtype: Float.tensorFlowDataType,
                                    strides: [1, 1, 1, 1],
                                    padding: loadVALIDString())
```

Can we assign `UnsafePointer<UInt8>` to type of `padding`, so that SILGen will emit the right code to do the conversion from the output of `loadVALIDString()` to `UnsafePointer<UInt8>`.

IRGen will then only deal with c arrays in the case of string attrs, and not Swift Strings.

Reply:
I'm not sure I got your first suggestion about reducing the magic. Are you suggesting that I make the swift function something like this?
````
func _TFCOpSetAttrString(..., ptr: UnsafePointer<Int8>, number: Int64) {
  <do some stuff to make a c string from the ptr and number that we have>
  TFE_OpSetAttrString(..., cstr)
}
````

I think that would couple things even more with the internal implementation details of String, because I would have to figure out what the values inside the String actually mean and how to get the data out of them. So I'd rather keep the magic llvm value -> swift value conversion that happens. As far as I can tell, it's not really that magical -- it's just the standard way that swift function calls get irgen'd to llvm function calls.

Reply:
discussed in person, resolved in  ac18326